backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/query/impl/LuceneChildrenCollector.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.search.query.impl;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.hibernate.search.backend.lucene.util.impl.LuceneFields;
+
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Scorable;
+import org.apache.lucene.search.ScoreMode;
+
+public class LuceneChildrenCollector implements Collector {
+
+	private final Map<String, Set<Integer>> children = new HashMap<>();
+	private static final Set<String> ROOT_ID_FIELD_SET = Collections.singleton( LuceneFields.rootIdFieldName() );
+
+	@Override
+	public LeafCollector getLeafCollector(LeafReaderContext context) throws IOException {
+		return new FieldLeafCollector( context );
+	}
+
+	@Override
+	public ScoreMode scoreMode() {
+		return ScoreMode.COMPLETE_NO_SCORES;
+	}
+
+	public Map<String, Set<Integer>> getChildren() {
+		return children;
+	}
+
+	private class FieldLeafCollector implements LeafCollector {
+
+		private final LeafReader reader;
+
+		public FieldLeafCollector(LeafReaderContext context) throws IOException {
+			reader = context.reader();
+		}
+
+		@Override
+		public void setScorer(Scorable scorer) throws IOException {
+			// we don't need any scorer
+		}
+
+		@Override
+		public void collect(int doc) throws IOException {
+			// FIXME tried to use BinaryDocValues without success. The field isn't nether binary nor numeric.
```

Comment:
It's a good idea to use doc values here. Currently we don't store doc values for this field, which is why BinaryDocValues didn't work, but we definitely should.

In `org.hibernate.search.backend.lucene.document.impl.LuceneNestedObjectDocumentBuilder#contribute`, you will have to use something else than a `StringField` for the "rootId" field.

You can just use `Field` instead of `StringField`, and pass your own field type, defined as a private static constant. See in `StringField` how field types are constructed.

Then doc values should work as expected. See `org.hibernate.search.backend.lucene.search.extraction.impl.DistanceCollector.DistanceLeafCollector` to get an idea of how doc values are used (it's not the same type of doc values, but usage should be similar).

Reply:
Very clear! Thank you for the explanation